{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2d310673",
   "metadata": {},
   "source": [
    "Here, my goal is to fit the model \n",
    "\n",
    "temp ~ P_embeds \n",
    "\n",
    "I will prepare P_embeds a bit like an alignment supermatrix; for a gene, rbcL, I will take all the AA sequences, collect the embeddings, which will be of dimension embed_dim,L+2, and get the embedding_dim mean, so that I have n_sample mean embedding vectors of length L + 2
    
    I will now make the data structure that will hold the esm embeddings
    Basically, it will be 
    ID Taxonomy Gene AA_seq embedding_1, embedding_2, 
    
    First thing I got to do - I have all the protein sequences for my species in data/tmp/genesToAlign/*AA.fasta. I can grab these guys, concatenate them, and ship them over to cbsugpu08
    
    
    The fasta headers are like ID|Taxonomy|Gene, and the fasta holds the AA_seq
    
    > ID|Taxonomy|Gene
    MTAILERS... 
    
  
    
    "
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
