---
title: "treee"
output: html_notebook
editor_options: 
  chunk_output_type: inline
---
```{r setup, include=FALSE}
# Load knitr
library(knitr)
library(ape)
library(data.table)
library(arrow)
library(glmnet)
library(doParallel)
library(ggplot2)
# Set the working directory for all chunks
opts_knit$set(root.dir = "/local/workdir/hdd29/chloroplast_genome_evaluation")
```



```{r}
#read in SNPs
aln <- read_parquet("data/tmp/majMinor_aln.pq")
setDT(aln)
for (j in seq_along(aln)) set(aln, which(is.na(aln[[j]])), j, median(aln[[j]], na.rm=TRUE))
#drop all zero var columns 
invariant_cols <- names(aln)[sapply(aln, function(x) length(unique(x)) == 1)]
aln <- aln[, !..invariant_cols]

#read in data
data <- read_parquet("data/processed_data.parquet")

dim(aln)
dim(data)
```



```{r}
cols <- c("firebrick","steelblue","seagreen","goldenrod")
names(cols) <- c("BIO8","GC","GenomeLength","GeneCount")

cv <- function(x) sd(x, na.rm=TRUE)/mean(x, na.rm=TRUE)
vals <- list(
  BIO8=data$pheno_wc2.1_2.5m_bio_8_p50,
  GC=data$geno_genomicGC,
  GenomeLength=data$geno_genomeLength,
  GeneCount=data$geno_geneCount
)

pdf("figures/hists_cv_vertical.pdf", width=6, height=10)
par(mfrow=c(5,1), mar=c(4,4,2,1))

for(n in names(vals)){
  hist(vals[[n]], breaks=50, main=n, xlab="", col=cols[n], border="white")
}

barplot(sapply(vals, cv), col=cols, las=2, ylab="CV", main="Coefficient of Variation")
dev.off()


```

```{r}
X <- as.matrix(aln[, grep("cds_supermatrix", colnames(aln), value=TRUE), with=FALSE])
y <- data$pheno_wc2.1_2.5m_bio_8_p50
```

```{r}
registerDoParallel(10)
set.seed(1)
n <- nrow(X)
test_idx <- sample(seq_len(n), size = floor(0.1*n))
train_idx <- setdiff(seq_len(n), test_idx)

cvfit <- cv.glmnet(X[train_idx,], y[train_idx], alpha=1, parallel = T, trace.it = 1)
fit <- glmnet(X[train_idx,], y[train_idx], alpha=1, lambda=cvfit$lambda.min)
```


```{r}
pred_train <- predict(fit, X[train_idx,])
pred_test  <- predict(fit, X[test_idx,])

r2_train <- 1 - sum((y[train_idx]-pred_train)^2)/sum((y[train_idx]-mean(y[train_idx]))^2)
r_test  <- cor(pred_test, y[test_idx])
```


```{r}
pdf("figures/1.pdf", width = 12, height=6)
par(mfrow=c(1,2))
plot(pred_train, y[train_idx], xlab="Predicted", ylab="Observed", main=paste("Train R2 =", round(r2_train,2),"\n", "n =", length(train_idx)))
abline(0,1,col="red")

plot(pred_test, y[test_idx], xlab="Predicted", ylab="Observed", main=paste("Random 10% Holdout Test pearson r =", round(r_test,2),"\n", "n =", length(test_idx)))
abline(0,1,col="red")
dev.off()
```

```{r}
# hold out Poales clade
clade_idx <- grep("Poales", data$Taxonomy)
train_idx2 <- setdiff(seq_len(n), clade_idx)
test_idx2  <- clade_idx

cvfit2 <- cv.glmnet(X[train_idx2,], y[train_idx2], alpha=1, parallel=TRUE)
fit2 <- glmnet(X[train_idx2,], y[train_idx2], alpha=1, lambda=cvfit2$lambda.min)

pred_clade <- predict(fit2, X[test_idx2,])
r_clade <- cor(pred_clade, y[test_idx2])
```

```{r}
pdf("figures/1.pdf", width=9, height=3)
par(mfrow=c(1,3))

plot(pred_train, y[train_idx], xlab="Predicted", ylab="Observed",
     main=paste("Train R2 =", round(r2_train,2),"\n", "n =", length(train_idx)))
abline(0,1,col="red")

plot(pred_test, y[test_idx], xlab="Predicted", ylab="Observed",
     main=paste("Random 10% Holdout r =", round(r_test,2),"\n", "n =", length(test_idx)))
abline(0,1,col="red")

plot(pred_clade, y[test_idx2], xlab="Predicted", ylab="Observed",
     main=paste("Poales Holdout r =", round(r_clade,2),"\n", "n =", length(test_idx2)))
abline(0,1,col="red")

dev.off()
```



The correlogram takes a bit of time, since we have to compute Moran's I over a large number of bins
from analyses/neighbourhood_crlgm.r
```{r}
moran.results <- readRDS("results/spermatophyta_neighborhood_moran_results.rds")
# extract values
I <- sapply(moran.results, function(x) {
  if (length(x) == 1 && is.na(x)) return(NA)
  x$obs[1]
})

pval <- sapply(moran.results, function(x) {
  if (length(x) == 1 && is.na(x)) return(NA)
  x$pvalue[1]
})

classes <- 1:length(I)

pdf("figures/spermatophyta_neighborhood_correlogram.pdf",
    width = 10, height = 5)
plot(classes, I, type = "b", pch = 16,
     main="Phylogenetic signal for BIO8",
     cex.main = 0.9,font.main = 2,
     xlab = "topological distance (nodes apart)", ylab = "Moran's I (Abouheif)")
abline(h = 0, lty = 2)
points(classes, I, col = ifelse(pval <= 0.01, "red", "black"))
legend("topright", legend = c("p < 0.01"), pch = 16, col = "red", bty = "n")
dev.off()
```



```{r}
# load tree + data
tree <- read.tree("data/2_global_order_level.tre")
tree$tip.label <- gsub("\\s+", "", tree$tip.label)
data$Order <- gsub("\\s+", "", as.character(data$Order))

keep <- intersect(tree$tip.label, data$Order)
tree <- keep.tip(tree, keep)
data$Order <- factor(data$Order, levels=tree$tip.label)

pdf("figures/tree_boxplot.pdf", width=12, height=8)
par(mfrow=c(1,5))
plot(tree, cex=0.6, main="Order level Tree")
boxplot(pheno_wc2.1_2.5m_bio_8_p50 ~ Order, data=data,
        horizontal=TRUE, las=1, outline=FALSE,
        xlab="BIO8 temperature", main="Trait distribution by Order",
        ylab="")
boxplot(data$geno_genomicGC ~ Order, data=data,
        horizontal=TRUE, las=1, outline=FALSE,
        xlab="GC Proportion", main="GC distribution",
        ylab="", yaxt="n")
boxplot(data$geno_genomeLength ~ Order, data=data,
        horizontal=TRUE, las=1, outline=FALSE,
        xlab="Genome Length (bp)", main="Genome Length",
        ylab="", yaxt="n")
boxplot(data$geno_geneCount ~ Order, data=data,
        horizontal=TRUE, las=1, outline=FALSE,
        xlab="Gene Count", main="Gene Count",
        ylab="", yaxt="n")
axis(4, at=1:length(levels(data$Order)), labels=levels(data$Order), las=1, cex.axis=0.6)
dev.off()

```

```{r}
traits <- c("pheno_wc2.1_2.5m_bio_8_p50","geno_genomicGC","geno_genomeLength","geno_geneCount")
cv_fun <- function(x) sd(x,na.rm=TRUE)/mean(x,na.rm=TRUE)
cv_df <- do.call(rbind, lapply(levels(data$Order), function(o) {
  sub <- subset(data, Order==o)
  c(Order=o, sapply(traits, function(t) cv_fun(sub[[t]])))
}))
cv_df <- as.data.frame(cv_df)
cv_df[,traits] <- lapply(cv_df[,traits], as.numeric)

n_df <- as.data.frame(table(data$Order))
colnames(n_df) <- c("Order","n")
cv_df$n <- n_df$n[match(cv_df$Order, n_df$Order)]

mat <- t(as.matrix(cv_df[,traits]))
barplot(mat, beside=TRUE, las=2, legend.text=traits, cex.names=0.6,
                      names.arg=cv_df$Order,
        args.legend=list(x="topright", bty="n"),
        main="CV by variable per order")

```




```{r}
pdf("figures/tree_boxplot.pdf", width=12, height=8)
par(mfrow=c(1,5))

plot(tree, cex=0.6, main="Order level Tree")

boxplot(pheno_wc2.1_2.5m_bio_8_p50 ~ Order, data=data,
        horizontal=TRUE, outline=FALSE,
        xlab="BIO8 temperature", ylab="", yaxt="n")

boxplot(geno_genomicGC ~ Order, data=data,
        horizontal=TRUE, outline=FALSE,
        xlab="Genomic GC proportion", ylab="", yaxt="n")

boxplot(geno_genomeLength ~ Order, data=data,
        horizontal=TRUE, outline=FALSE,
        xlab="Genome length (bp)", ylab="", yaxt="n")

boxplot(geno_geneCount ~ Order, data=data,
        horizontal=TRUE, outline=FALSE,
        xlab="Gene count", ylab="", yaxt="n")

axis(4, at=1:length(levels(data$Order)), labels=levels(data$Order), las=1, cex.axis=0.6)

dev.off()


```


```{r}
embeds <- readRDS("data/tmp/embeds_with_mds.rds")
df <- read_parquet("data/processed_data.parquet")
setDT(embeds); setDT(df)

clean_embeds <- embeds[ManualOutlier==FALSE]
embed_cols <- grep("embedding", colnames(embeds), value=TRUE)
pheno_col <- "pheno_wc2.1_2.5m_bio_8_p50"
genes <- unique(clean_embeds$Gene)

merged_list <- vector("list", length(genes))
names(merged_list) <- genes

for (gene in genes) {
  gdt <- clean_embeds[Gene==gene, c("ID", embed_cols), with=FALSE]
  new_names <- paste0(gene, "__", embed_cols)
  setnames(gdt, old=embed_cols, new=new_names)
  merged_list[[gene]] <- gdt
}

X_dt <- Reduce(function(a,b) merge(a,b,by="ID",all=TRUE), merged_list)
tax <- unique(clean_embeds[, .(ID, Taxonomy)])
X_dt <- merge(X_dt, tax, by="ID", all.x=TRUE)

pred_cols <- setdiff(colnames(X_dt), c("ID","Taxonomy"))
#for (col in pred_cols) {
#  med <- median(X_dt[[col]], na.rm=TRUE)
#  X_dt[is.na(get(col)), (col):=med]
#}
saveRDS(X_dt, "data/tmp/x_dt_imputed.rds")
```
```{r}
X_dt <- readRDS("data/tmp/x_dt_imputed.rds")
X <- as.matrix(X_dt[, ..pred_cols])
y <- df[match(X_dt$ID, df$ID), get(pheno_col)]
n <- nrow(X)
```


```{r}
registerDoParallel(10)
set.seed(1)

test_idx <- sample(seq_len(n), floor(0.1*n))
train_idx <- setdiff(seq_len(n), test_idx)

cvfit <- cv.glmnet(X[train_idx,], y[train_idx], alpha=1, parallel=TRUE, trace.it=1)
fit <- glmnet(X[train_idx,], y[train_idx], alpha=1, lambda=cvfit$lambda.min)

pred_train <- predict(fit, X[train_idx,])
pred_test <- predict(fit, X[test_idx,])

r2_train <- 1 - sum((y[train_idx]-pred_train)^2)/sum((y[train_idx]-mean(y[train_idx]))^2)
r_test <- cor(pred_test, y[test_idx])
```

```{r}
clade_ids <- df[grepl("Poales", Taxonomy), ID]

clade_idx <- which(X_dt$ID %in% clade_ids)
train_idx2 <- setdiff(seq_len(n), clade_idx)
test_idx2 <- clade_idx

cvfit2 <- cv.glmnet(X[train_idx2,], y[train_idx2], alpha=1, parallel=TRUE, trace.it=1)
fit2 <- glmnet(X[train_idx2,], y[train_idx2], alpha=1, lambda=cvfit2$lambda.min)

pred_clade <- predict(fit2, X[test_idx2,])
r_clade <- cor(pred_clade, y[test_idx2])

```

```{r}
pdf("figures/embeds_lasso_baseline.pdf", width=9, height=3)
par(mfrow=c(1,3))

plot(pred_train, y[train_idx],
     xlab="Predicted", ylab="Observed",
     main=paste("Train RÂ² =", round(r2_train,2), "\n","n =", length(train_idx)))
abline(0,1,col="red")

plot(pred_test, y[test_idx],
     xlab="Predicted", ylab="Observed",
     main=paste("Random 10% Holdout r =", round(r_test,2), "\n","n =", length(test_idx)))
abline(0,1,col="red")

plot(pred_clade, y[test_idx2],
     xlab="Predicted", ylab="Observed",
     main=paste("Poales Holdout r =", round(r_clade,2), "\n","n =", length(test_idx2)))
abline(0,1,col="red")

dev.off()
```


